File: test-suite/src/test/java/io/micronaut/docs/http/client/bind/annotation/MetadataClientArgumentBinder.java
package io.micronaut.docs.http.client.bind.annotation;

//tag::clazz[]
import io.micronaut.core.annotation.NonNull;
import io.micronaut.core.convert.ArgumentConversionContext;
import io.micronaut.core.naming.NameUtils;
import io.micronaut.core.util.StringUtils;
import io.micronaut.http.MutableHttpRequest;
import io.micronaut.http.client.bind.AnnotatedClientArgumentRequestBinder;
import io.micronaut.http.client.bind.ClientRequestUriContext;

import jakarta.inject.Singleton;
import java.util.Map;

@Singleton
public class MetadataClientArgumentBinder implements AnnotatedClientArgumentRequestBinder<Metadata> {

    @NonNull
    @Override
    public Class<Metadata> getAnnotationType() {
        return Metadata.class;
    }

    @Override
    public void bind(@NonNull ArgumentConversionContext<Object> context,
                     @NonNull ClientRequestUriContext uriContext,
                     @NonNull Object value,
                     @NonNull MutableHttpRequest<?> request) {
        if (value instanceof Map map) {
            for (Map.Entry<?, ?> entry: map.entrySet()) {
                String key = NameUtils.hyphenate(StringUtils.capitalize(entry.getKey().toString()), false);
                request.header("X-Metadata-" + key, entry.getValue().toString());
            }
        }
    }
}
//end::clazz[]

File: test-suite-python/src/test/python/micronaut/docs/http/client/bind/annotation/MetadataClientArgumentBinder.py
import java

# tag::clazz[]
from jakarta.inject import Singleton
from micronaut.core.convert import ArgumentConversionContext
from micronaut.core.naming import NameUtils
from micronaut.core.util import StringUtils
from micronaut.http import MutableHttpRequest
from micronaut.http.client.bind import AnnotatedClientArgumentRequestBinder, ClientRequestUriContext

Class = java.type("java.lang.Class")
MetadataClass = java.type("micronaut.docs.http.client.bind.annotation.Metadata")


@Singleton
class MetadataClientArgumentBinder(AnnotatedClientArgumentRequestBinder):

    def getAnnotationType(self) -> Class:
        return MetadataClass

    def bind(
        self,
        context: ArgumentConversionContext,
        uriContext: ClientRequestUriContext,
        value: object,
        request: MutableHttpRequest,
    ) -> None:
        if hasattr(value, "entrySet"):
            for entry in value.entrySet():
                key = NameUtils.hyphenate(StringUtils.capitalize(str(entry.getKey())), False)
                request.header("X-Metadata-" + key, str(entry.getValue()))
# end::clazz[]
